### PR TITLE
Add legal hold membership to device reporting

### PR DESCRIPTION
**Pull request by maddie-vargo**
**Description of Change**

This PR addresses Phase I of #176 to add an `--include-devices` option to the legal-hold command to show devices associated with the legal hold custodians captured by the query. This option will also total storage held by these devices and print it in a table grouped by organization. 

This PR requires pandas as a dependency. Python 3.5 does not support the pandas library, so this should not be merged until support ends for 3.5.

**Testing Procedure**

```
code42 legal-hold show 1234 --include-devices
```

**PR Checklist**
Did you remember to do the below?

- [x]  Add unit tests to verify this change
- [x]  Add an entry to CHANGELOG.md describing this change
- [x]  Add docstrings for any new public parameters / methods / classes

### CODE REVIEW

**Comment by maddie-vargo** [comment]
I have read the CLA Document and I hereby sign the CLA

File: CHANGELOG.md
Comment (unparalleled-js) [inline CHANGELOG.md:37]: We do have a new `devices list` command that is merged by not yet released.

Would it make more sense to have a command `code42 devices list --legal-hold-uid <UID>` to instead list? Basically a command that filters devices based on a legal-hold uid.

Just want to make sure all the options are considered, I don't really understand the use-case.

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:140]: can just be `if len(devices_dataframe.index)`

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:142]: We typically use table format for show commands.

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py:142]: List commands, however, are known to support various output formats, inc. CSV.... My first comment on this PR suggests that this would make sense as a `code42 devices list --matter-id <UID>` command, which then you can tack on `-f CSV` or `-f JSON` etc.... Would that be better?

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:267]: py42 prefers a bool for the `active` param

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py:267]: If for some reason the request fails in py42 when you do that, then this is fine, but I'd like to open a PR for py42 there if that is the case

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:298]: can we call this var something other than `i`?

File: src/code42cli/cmds/legal_hold.py
Comment (unparalleled-js) [inline src/code42cli/cmds/legal_hold.py:304]: This method does seem to actually print any storage by org

File: tests/cmds/test_legal_hold.py
Comment (unparalleled-js) [inline tests/cmds/test_legal_hold.py:464]: I find it is better to actually use generators for mocks when they are supposed to be rather than lists.. It saved future headaches, in my experience.

Comment (unparalleled-js) [reply tests/cmds/test_legal_hold.py:464]: Might need to use `side_effect` instead of `return_value` in that case, not sure

File: src/code42cli/cmds/legal_hold.py
Comment (maddie-vargo) [reply src/code42cli/cmds/legal_hold.py:304]: I'm not sure I understand the comment. Do you mean _doesn't_ actually print storage by org? If so, you're right. It just _gets_ it.

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py]: The name of the method suggests it prints it

Comment (unparalleled-js) [reply src/code42cli/cmds/legal_hold.py]: I miswrote the original comment

File: CHANGELOG.md
Comment (unparalleled-js) [reply CHANGELOG.md:37]: Or a `code42 legal-hold list-devices [MATTER-ID]`.  Trying to find a way to afford  output format options (At least table and csv).

Comment (maddie-vargo) [reply CHANGELOG.md:37]: I agree, if we can find a way to structure this to add a format option, that would be helpful to most customers. 

Currently, the `legal-hold` show command outputs data in the following formats:

* Matter info - table
* Users - columns
* (optional) Preservation policy - JSON


What do you think about removing users from the `show` command and creating `code42 legal-hold list-users [Matter ID]` with an `--include-devices` flag? I think it makes sense to keep users and devices closely tied together, rather than in separate commands.

Comment (unparalleled-js) [reply CHANGELOG.md:37]: I like that! @timabrmsn What are you thoughts on this?  I believe you were the one who implemented the show command.

Comment (timabrmsn) [reply CHANGELOG.md:37]: I actually think it makes more sense to add legal hold status as an option to the `devices list` command, since that would be a useful option for people who are creating a report of all their devices to know which ones are on legal hold quickly when taking inventory of other things. It could just add a `legalHoldMatterId` column to the output. 

And keeping users under `legal-hold show` still makes sense to me, since it's a high level overview of the matter, of which membership is a key detail. Otherwise the `show` command just becomes a single row of the `list` command with the option to also show preservation policy. 

`devices list` already has `--include-usernames` which makes a separate call to `api/Users` and then merges the DataFrames on `userId`, so adding legal hold status would be easy enough to add using the same logic.

The storage total report is an interesting idea, but might be better suited to include in `devices list` report also, so it can be used more generically and not tied directly to only legal hold. I'm thinking something like `--include-storage-totals` on `devices list` that just adds a column per device for the sum of all that device's archives. Then if someone wanted to get a report broken down by matter Id, they could just filter by matter_id and sum the total in the csv report.

Comment (maddie-vargo) [reply CHANGELOG.md:37]: Ok, `devices` it is. The device pull actually includes `userUid` without the username option, so it's even easier to merge in the legal hold data. 

There are two main things that we want this to accomplish: 

* Limit the device list to devices on a specific legal hold matter
`devices --matter-uid` - option to limit device search base on a specific legal hold matter, adds in columns such as legalHoldStatus, legalHoldMatterId, legalHoldMatterName 

* Add in legal hold membership columns for whatever device query you're running. 
 `devices -include-legal-hold-membership` - option to option to include legal hold membership on to the devices table, adds same columns as above but doesn't limit the population

This feels a little redundant, so any feedback is appreciated. 

---

As for `storage-totals`: is a sum of all device's archives appropriate? (this is how I wrote my first draft). However, the current legal hold script does some logic to determine which destination is the most recent and complete, and only uses that archive size as the total storage.

Comment (timabrmsn) [reply CHANGELOG.md:37]: I think that since filtering by matter-id outside of the CLI is simple enough (i.e. grep/Select-String or directly in excel), can we leave that to the end-user? I'd like to not have our option list expand too much. And I have some ideas for more generic column/row filtering logic that could work on any cmd that outputs a dataframe. Something like `--filter 'matter-id == 42' that would convert to a pandas expression under the hood. Although I haven't started trying to implement that so I'm not sure how feasible it is...

Also, because a device can be part of multiple legal holds, we need to think of how to display that. Can we just have a single column that has a list of matter-uids the device is part of, maybe separated by semi-colons or something?

Comment (timabrmsn) [reply CHANGELOG.md:37]: Yeah, for storage totals I think adding a `usedStorage` column that sums the total combined archive size for a device would work, maybe we could also add an `archiveCount` column that gets added with that option?

Trying to keep what features we add as generic and multi-use-case friendly as possible.

**Comment by maddie-vargo** [comment]
I have committed new changes that remove the edits to `legal-hold` and move it all to the `devices` command.

- If a device is on multiple legal holds, I concatenate the IDs, names, and status into a comma-separated list. 
- Since legal hold membership is determined at the user level, deactivated devices can print legal hold membership. It did not seem meaningful to present that for deactivated devices, so I cleared those cells.

As for the `used-storage` piece - this data is present if the user includes the `--include-backup-usage` option. In efforts to keep this as generalized as possible, is it too much to add in a `--include-total-storage` option that would bring in a column that totals device storage across destination? Re: Tim's comment above, this calculation could be left to the end user.

**Comment by kiran-chaudhary** [comment]
Can we update the description please based on latest changes. I thought to try it out reading the description and then looking at the code it confused me more.

File: src/code42cli/cmds/devices.py
Comment (kiran-chaudhary) [inline src/code42cli/cmds/devices.py:16]: methods staring with `_` represent private methods, so if these methods are needed feel free to make them public.

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:355]: I agree with Kiran above that we probably shouldn't use the legal-hold module's internal functions, just for the sake of if someone later needs to change them it could break stuff here. 

Since we only need to iterate through all membership dicts, we can simplify getting them all with a generator function instead of building a new list and extending it: 

```
def _get_all_active_hold_memberships(sdk):
    for page in sdk.legalhold.get_all_matters(active=True):
        for matter in page["legalHolds"]:
            for _page in sdk.legalhold.get_all_matter_custodians(
                legal_hold_uid=matter["legalHoldUid"], active=True
            ):
                for membership in _page["legalHoldMemberships"]:
                    yield membership
```

And pandas has a great function that can 'flatten' nested json into a dataframe for you, which can simplify a lot of this logic: 

```
>>> df = pandas.json_normalize(_get_all_active_hold_memberships(sdk))
>>> df.head()
  legalHoldMembershipUid  active                   creationDate legalHold.legalHoldUid legalHold.name        user.userUid                   user.username                      user.email user.userExtRef
0     955061003769628092    True  2020-05-20T15:58:38.823-05:00     955060991270602172          Test2  945056771151950748                test@example.com                test@example.com            None
1     988814086101244865    True  2021-01-08T11:25:23.645-06:00     946193444244471969           Test  973619347957360307                legalhold@code42.com            legalhold@code42.com        None
2     955320624671252917    True  2020-05-22T10:57:44.903-05:00     946193444244471969           Test  955163198918558214                de60@example.com                de60@example.com            None
3     955320624637698485    True  2020-05-22T10:57:44.892-05:00     946193444244471969           Test  955163197341499910                de59@example.com                de59@example.com            None
4     955320624604144053    True  2020-05-22T10:57:44.885-05:00     946193444244471969           Test  955163195865104902                de58@example.com                de58@example.com            None
```

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:395]: Instead of using np.nan for null values here, let's use an empty string, I think `NaN` would confuse people in the output and think maybe there was a bug. To replace any np.nan values resulting from the `.groupby()`, you can call `.fillna(value="")` on the df to replace those as well.

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:388]: You can just pass `",".join` here instead of wrapping it in a lambda.

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:362]: I changed it in my example `_get_all_active_hold_memberships` above, but does it make sense to include inactive memberships in these results? If we only get active memberships, then we can drop the `legalHoldMemberActive ` column from the results and the data will read cleaner.

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py:16]: I'll resolve this by moving towards Tim's suggestion below

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py:395]: I chose `NaN` to match the value assumed by pandas of an empty value. `NaN`'s are generated when the legal hold table is merged with the devices table. If I use the above solution prior to merging with the devices table, the legalHoldUid and legalHold.name columns will contain a mix of None's, actual values, and NaNs.

To confirm: which of the following would you prefer?

* NaN's in table view
* None's in table view

(Either will produce null values in the other output formats)

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py:362]: I had included inactive memberships as a relic from when this code lied in the `legal-hold` command, where there is an `--include-inactive` option. I checked with Paul Hirst, and there's not much of a use case for this, so I'll go with your suggestion.

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py:355]: `json_normalize` is great for unpacking the nested dictionary. However, I seem to be losing records (1 per matter). If I run the generator response table through `pandas.DataFrame`, I get a table with more rows than I do running the same generator response through `json_normalize`. I haven't quite worked this out yet

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [reply src/code42cli/cmds/devices.py:395]: Oops, I meant to say "resulting from the`.merge()`" instead of `.groupby()`. Yeah, the `.fillna(value="")` needs to be called after the merge. 

I think we should convert NaN/None values to an empty string for table/CSV outputs, but we'd probably want to keep them as NaN/None when we output to json so it translates to `null`. But we should push that logic to the `DataFrameOutputFormatter` so we only need to handle it in one place for any commands that use DataFrames.

I'll start a new PR to rework the formatter to handle that, since I'd like to refactor it a bit for clarity and add some tests around handling dataframe nulls.

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [reply src/code42cli/cmds/devices.py:355]: Interesting, I'll see if I get the same behavior and try to figure out what's happening too.

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py]: NaNs removed

File: src/code42cli/cmds/devices.py
Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py]: I had to serialize the generator into a list in order for all records to be included, which may defeat the purpose of the generator.

Comment (timabrmsn) [reply src/code42cli/cmds/devices.py]: Looks like there was a recent pandas bug about the generator thing (they weren't expecting generators to be passed): https://github.com/pandas-dev/pandas/issues/35923

It's fixed in 1.2.2, but I think we can probably just call list() on the result for now instead of force updating to the latest pandas.

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [reply src/code42cli/cmds/devices.py]: Ok, I've opened https://github.com/code42/code42cli/pull/245 to add the null handling in the DataFrameOutputFormatter, so we don't need to be concerned with converting them here, since we want the conversion to be conditional on the output format chosen by the end-user. 

Sorry to have you flip/flop this!

Comment (maddie-vargo) [reply src/code42cli/cmds/devices.py:370]: Resolved, and awaiting #245

**Comment by maddie-vargo** [comment]
All PR comments should be resolved. 

The PR does not include a per-org storage calculation. The `--include-backup-usage` options reports archive size per archive. Is it preferred to create another option for `--include-storage-totals`, or rely on the user to calculate these totals outside of the CLI?

**Comment by timabrmsn** [comment]
Hmm, it seems that the `--include-backup-usage` logic was broken somewhere, since that option now doesn't actually add anything to the output..

**Comment by maddie-vargo** [comment]
I'm not able to recreate that behavior - the option spits out backup usage data on my branch. Diffing with the upstream code42cli master doesn't show any changes to backup usage logic

**Comment by timabrmsn** [comment]
Nope, it was just a device with a _really_ large backup usage result towards the bottom of the list that was pushing that column way to the right.

I think a separate option + column for total storage a device is using would be really helpful. I also am thinking `--include-backup-usage` would probably be better turned into a separate command like `list-archives`, where it's one row per archive like our `list-backup-sets`. Right now parsing any specific data out of the `backupUsage` column is difficult for our end-users.

But that can be a separate PR...

I know a big part of the goal of this PR is to enable legal hold storage reporting, right? If we add an `--include-storage-totals` option that ensures `include_backup_usage` is True on the api call (but not necessarily on the table output), and then extracts and sums the total archiveBytes for each device and sticks it in a `totalUsedStorage` column, is that close enough for your customers who want these reports?

**Comment by timabrmsn** [comment]
@maddie-vargo PR #245 just merged

**Comment by maddie-vargo** [comment]
`--include-total-storage` option now pushed up

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:468]: I wonder if we should filter out any V2 archives from these results, as it could cause confusion if 2 archives show up in the list when only 1 is visible in console. We can clarify this in the cmd help text to say `Include backup archive count and total storage in output.`

**Comment by timabrmsn** [comment]
I think this is good, just the small thing of filtering out V2 archives from the storage totals. I'll write up a ticket for testing with all the acceptance criteria and make sure I haven't missed anything else as I go through documenting each change.

**Comment by timabrmsn** [comment]
`json_normalize` names the unpacked nested columns with dotted notation `legalHold.legalHoldUid`, `legalHold.name` can we rename this to just `legalHoldUid` and `legalHoldName`?

File: src/code42cli/cmds/devices.py
Comment (timabrmsn) [inline src/code42cli/cmds/devices.py:474]: This should be `if archive["archiveFormat"] != "ARCHIVE_V2":`, since while V3 archives are not prevalent, they still should show up in the results.

File: CHANGELOG.md
Comment (kiran-chaudhary) [inline CHANGELOG.md:78]: can we rephrase `printing ....` to something like `prints the legalhold matter name and ID for any active device on legal hold`!? 

atleast `active device actively` seems off while reading.

**Review by kiran-chaudhary — APPROVED** [review]
Apart from `change-log` comment LGTM.

File: CHANGELOG.md
Comment (maddie-vargo) [reply CHANGELOG.md]: yes, that is terrible phrasing. Resolved with latest commit

File: CHANGELOG.md
Comment (unparalleled-js) [inline CHANGELOG.md:61]: Looks like our changelog has gotten messed up? The ` #Unreleased` tag should not be there...

Comment (maddie-vargo) [reply CHANGELOG.md]: @unparalleled-js I re-pulled from master. The #Unreleased tag should be at the top, right? Let me know if I should fix it with this PR

**Review by timabrmsn — APPROVED** [review]

**Pull request merged**
